### ISSUE TITLE
optimization with optim package

### ISSUE THREAD

**Issue by chenwangliangguo** [harvardnlp/seq2seq-attn#34]
Hi ,great job. 
It seems that the optimization part is done manually , i wonder whether can we combine optim package (or some other automatic way)with it or not, if yes, how to do that, i am a new comer to torch, thank you in advance.

**Comment by yoonkim** [comment]
This is going to be a little annoying to do because the optim function expects as one of its inputs, a function that calculates the loss and returns the derivative (as a single, one-dim tensor). However, we have three tensors of derivatives (encoder, decoder, and linear+softmax--what i call the generator).

Hope this helps!

**Comment by chenwangliangguo** [comment]
@yoonkim   Thank you for your reply.
another issue: what is the meaning of function of setReuse, why set the gradInput equal to the the  output, in my opinion, they are not the same thing:

        function nn.Module:setReuse()
            if self.reuse then
             self.gradInput = self.output
          end
       end

**Comment by yoonkim** [comment]
That's just a hack to share memory between gradinput/output for certain layers.
This reduces memory footprint somewhat.
